Migrates GlobalRoles and GlobalRoleBinding.

KevinJoiner opened the pull request:
Migrates Noramn validation done on GlobalRoles and GlobalRoleBinding in Rancher to the webhook.

## Problem
Norman validation and mutation for GlobalRoles and GlobalRoleBindings need to be migrated to the webhook

## Solution
### Global Role Validation added.
#### Builtin Validation

The `globalroles.builtin` field is immutable, and new builtIn GlobalRoles cannot be created.
If `globalroles.builtin` is true then all fields are immutable except  `metadata` and `newUserDefault`.
If `globalroles.builtin` is true then the GlobalRole can not be deleted.

### GlobalRoleBinding Validation and mutation added.

#### Invalid Fields - Update
Users cannot update the following fields after creation:
- `userName` 
- `groupPrincipalName`
- `globalRoleName`


#### Invalid Fields - Create
GlobalRoleBindings must have either `userName` or `groupPrincipalName`, but not both.

#### On create

When a GlobalRoleBinding is created an owner reference is attached to the backing GlobalRole defined by `globalRoleName`.

## CheckList
  <!-- 
  Test: 
   PRs should be accompanied by tests, even if there isn't a single test yet.  
   Unit tests are preferred over the addition of integration tests.
   If this PR does not require additional tests, state the reason below for reviewers.
  -->
- [x] Test
  <!-- 
  Docs: 
   If you are updating or creating a mutator or validator, you will also need to update or create the markdown that documents validator's or mutator's behavior.
   For more info on how docs work, see: https://github.com/rancher/webhook#docs
  -->
- [x] Docs

MbolotSuse commented on pkg/resources/management.cattle.io/v3/globalrolebinding/GlobalRoleBinding.md:
```diff
@@ -8,4 +8,21 @@ Users can only create/update GlobalRoleBindings with rights less than or equal t
+
+### On create
+
+When a GlobalRoleBinding is created an owner reference is attached to the backing GlobalRole defined by `globalRoleName`.
```
Right now it sounds like the owner ref is on the GlobalRole (not referring to the global role).

```suggestion
When a GlobalRoleBinding is created an owner reference is created on the binding referring to the backing GlobalRole defined by `globalRoleName`.
```

MbolotSuse reviewed (changes requested):
Still need to look at tests, but here's what I see on the prod-side code so far.

KevinJoiner commented on pkg/resources/management.cattle.io/v3/globalrole/validator.go:
```diff
@@ -75,23 +79,34 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
+			// This is needed to supported the deletion of old GlobalRoles that would not pass the update check that verifies all rules have verbs.
+			return admission.ResponseAllowed(), nil
+		}
+		if isMetaOnlyChange(oldGR, newGR) {
+			// if this change only affects metadata, don't validate any further
+			// this allows users with the appropriate permissions to manage labels/annotations/finalizers
+			return admission.ResponseAllowed(), nil
+		}
+		fieldErr := a.validateUpdateFields(oldGR, newGR, fldPath)
```
`isMetaDataOnly` is costly due to it's calls `DeepEqual` on metadata. `grContentEqual` alone is much cheaper. That being said, I did try out the change you suggested, but the code quality wasn't great since the second call ignores the `NewUserDefault` field.

MbolotSuse replied:
> isMetaDataOnly is costly due to it's calls DeepEqual on metadata. 

I'm not quite sure that I agree here - even though reflect isn't the quickest package, I would expect that the metadata (which on average is a smaller object) will be much quicker to compare than the rules fields (particularly at scale with large rules).

>  That being said, I did try out the change you suggested, but the code quality wasn't great since the second call ignores the NewUserDefault field.

Ok, I'll resolve this since IIRC even deep equals on everything takes quite a few rules before it starts to scale sufficiently badly (and I didn't have a super clean solution in mind).

KevinJoiner commented on pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go:
```diff
@@ -172,8 +182,16 @@ func (a *admitter) validateGlobalRole(globalRole *v3.GlobalRole, fieldPath *fiel
+// 2. Object is a metadata only change. (this allows users with the appropriate permissions to manage labels/annotations/finalizers)
+// deletionIsScheduled returns true if the GlobalRoleBinding is being deleted.
+func canSkipValidation(op admissionv1.Operation, oldGRB, newGRB *v3.GlobalRoleBinding) bool {
+	return op == admissionv1.Update && (newGRB.DeletionTimestamp != nil || isMetaOnlyChange(oldGRB, newGRB))
```
I agree that it is not necessary, but I liked the increased readability of having a separate function that explains the check.

KevinJoiner commented on pkg/resources/management.cattle.io/v3/globalrole/validator.go:
This was missed I updated the validation and tests.

KevinJoiner commented on pkg/resources/management.cattle.io/v3/globalrolebinding/main_test.go:
```diff
@@ -0,0 +1,339 @@
+	if test.args.oldGRB != nil {
+		oldGRB = test.args.oldGRB()
+	}
+	if oldGRB != nil {
+		req.Operation = v1.Update
+		req.Name = oldGRB.Name
+		req.Namespace = oldGRB.Namespace
+		req.OldObject.Raw, err = json.Marshal(oldGRB)
+		assert.NoError(t, err, "Failed to marshal GRB while creating request")
+	}
+	if newGRB != nil {
+		req.Name = newGRB.Name
+		req.Namespace = newGRB.Namespace
+		req.Object.Raw, err = json.Marshal(newGRB)
+		assert.NoError(t, err, "Failed to marshal GRB while creating request")
```
I don't think this can be combined since one is checking the function, and the other is checking the output of the function. If you can provide an example that doesn't call `test.args.newGRB()` twice I will update.

JonCrowther replied:
I hadn't considered the case where the function exists but it doesn't return a GRB. I think we can leave it as is then

KevinJoiner commented on pkg/resources/management.cattle.io/v3/globalrolebinding/mutator.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// MutatingWebhook returns the MutatingWebhook used for this CRD.
+func (m *Mutator) MutatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.MutatingWebhook {
```
I will test this for any errors.

tomleb commented on pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go:
```diff
@@ -172,8 +182,16 @@ func (a *admitter) validateGlobalRole(globalRole *v3.GlobalRole, fieldPath *fiel
 }
 
+// canSkipValidation checks if validation matches one of two scenarios which allow us to skip validation.
+// 1. Object is scheduled for deletion (deletion timestamp set)
```
Why do we allow updates if the object is pending deletion?

KevinJoiner replied:
So controllers can remove their finalizers since K8s will block deletes until the finalizers are removed. I will also note that this is safe since you can not remove the deletion timestamp once it has been set.

tomleb replied:
>  I will also note that this is safe since you can not remove the deletion timestamp once it has been set.

Yeah, I was wondering if there was a possibility that somehow (finalizers perhaps) a GRB stays alive (although still in pending delete state) and an update could still cause some security issues. eg: GRB binds to `regular-user`, then is set to be deleted but stays alive for a while, and is updated to bind to `superuser`. Does that make any sense?

KevinJoiner replied:
Yeah, that is a valid concern our controller, which assign permissions based on GRB, check if the deletion timestamp is set before operating on the object.

MbolotSuse reviewed (changes requested):
Couple of small comments on tests.

KevinJoiner commented on pkg/resources/management.cattle.io/v3/globalrolebinding/mutator_test.go:
```diff
@@ -0,0 +1,123 @@
+	validator := globalrolebinding.NewMutator(globalRoleCache)
+
+	tests := []testCase{
+		{
```
Good suggestion I found an edge case that was missed. I am unable to test a failed patch since an invalid request body get's caught during decode.

KevinJoiner commented on pkg/resources/management.cattle.io/v3/globalrolebinding/validator_test.go:
```diff
@@ -1,583 +1,591 @@
-		name                 string
-		globalRoleBinding    *v3.GlobalRoleBinding
-		oldGlobalRoleBinding *v3.GlobalRoleBinding
-		operation            admissionv1.Operation
```
Yeah, I agree these changes stemmed from the previous PR, which was written before any tests for GR and GRBs.

KevinJoiner commented:
1. Updated the Mutator to be idempotent. Check if the ref exist before adding it again.
2. added a reference to the struct changes in go.mod that should be removed when https://github.com/rancher/rancher/pull/42667 is merged.
3. Reverted dd7b7b1897a51dde8374554fa3ec332ea7566264

MbolotSuse commented on pkg/resources/management.cattle.io/v3/globalrolebinding/mutator.go:
```diff
@@ -0,0 +1,100 @@
+	}
+	newRefStr := ownerReference.String()
+	for i := range newGRB.OwnerReferences {
+		if newGRB.OwnerReferences[i].String() == newRefStr {
```
I'm a bit uncomfortable on relying on `.String()` producing a value suitable for equality checks. Put differently, how confident are we that `a.String() == b.String()` iff `a == b` here? If you have some examples from other projects using this that would probably help.

KevinJoiner replied:
Not tied to the usage of `String()`, I can also just check all the fields. I am confident  a.String() == b.String() iff a == b since the function is generated based on all of the fields [here](https://github.com/kubernetes/apimachinery/blob/v0.28.1/pkg/apis/meta/v1/generated.pb.go#L4610-L4624)

MbolotSuse replied:
I'd go with the value comparison, that seems more solid to me than the string comparison.

tomleb replied:
In case it can be useful, controller-runtimes implements a `SetOwnerReference()` function: https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/controller/controllerutil/controllerutil.go#L97. If you follow the code you see that they compare the group, the kind and the name. Not sure if we can use that function directly, maybe you know better here.

MbolotSuse reviewed (changes requested):
Mostly LGTM - leaving this comment more as a block until the fork is replaced.

One minor concern on the owner ref.

tomleb commented on pkg/resources/management.cattle.io/v3/globalrole/validator.go:
```diff
@@ -124,6 +139,30 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
+}
+
+// validateCreateFields blocks the creation of builtin globalRoles
+func validateCreateFields(oldRole *v3.GlobalRole, fldPath *field.Path) *field.Error {
```
When are built-in roles created and by who? How is it bypassing this check?

MbolotSuse replied:
Rancher creates the built-in GlobalRoles (as well as built-in role templates): https://github.com/rancher/rancher/blob/49e4618602de246cd51188d9338fb850ffac975e/pkg/data/management/rolebuilder.go#L318. To bypass the webhook, it uses the [bypass method](https://github.com/rancher/rancher-docs/blob/08ec8f2bddb89617c8f7cd0277f7649f8bbfb61a/docs/reference-guides/rancher-webhook.md#bypassing-the-webhook) (docs are being worked on), as you can see here: https://github.com/rancher/rancher/blob/49e4618602de246cd51188d9338fb850ffac975e/pkg/data/management/rolebuilder.go#L337.

tomleb replied:
Good to know

MbolotSuse commented on pkg/resources/management.cattle.io/v3/globalrolebinding/mutator.go:
```diff
@@ -0,0 +1,104 @@
+		UID:        globalRole.UID,
+	}
+	for i := range newGRB.OwnerReferences {
+		if newGRB.OwnerReferences[i].APIVersion == ownerReference.APIVersion &&
```
Personally speaking, I prefer @tomleb 's suggestion here: https://github.com/rancher/webhook/pull/288#discussion_r1319071063 about the use of an upstream library, but since they are effectively very similar I won't block the PR based on that.

KevinJoiner replied:
The use cases are a little different here. We are trying to keep our mutator idempotent, and this call only runs on create. The controller-runtime function is designed to run in a controller to reconcile the state of an object on change. I also don't like the behavior of overwriting the user's OwnerRef if they match GVK. Other nits are that the function is a lot heavier and more complex than our basic needs, and requires us to make a `runtime.Scheme`

MbolotSuse replied:
> and requires us to make a runtime.Scheme

Yeah that part isn't great, though IIRC you should be able to import a schemeBuilder from v3 and use that to get a workable scheme somewhat easily.

> I also don't like the behavior of overwriting the user's OwnerRef if they match GVK

I'm not sure that I agree with your assessment here, from what I can see it needs to match Group, Version, Kind, and Name - https://github.com/kubernetes-sigs/controller-runtime/blob/2a553d6f910dfc4b0796142ba1e059639a21c54a/pkg/controller/controllerutil/controllerutil.go#L170, meaning that it will only match if it's for the exact GR by name. UID should be the same for resources of the same GVK + Name, and the block owner deletion seems more like a user-set option and less like an exact equality match.

That being said, I get what you mean here, and like I said, not a blocking comment.

MbolotSuse replied:
Discussed offline, in order to not modify the user's input we are going to stay with the current implementation which never overrides the user's input.

MbolotSuse reviewed (approved):
One nit, lgtm outside of that.

KevinJoiner commented:
Rebased with with latest go.mod changes
